Taint analysis to improve the precision loss of partial contexts

FelixKrayer opened the pull request:
Adding a new analysis `taintPartialContexts` to track modified lvalues per function. The gained information is used in `base`, `relationAnalysis.apron`, `condVars` and `varEq` to keep unwritten values. (Issue #553)

The analysis `taintPartialContexts` tracks a set of modified values for each function, starting with the empty set when entering a function. This set can be accessed via a new query `MayBeTainted`, which is used in the combine of the four mentioned analyses to improve precision when partial contexts are used.

- `base`: the `cpa` from the caller is updated: 
	- everything is removed, which does not exist in the callee anymore
	- everything is added from callee, which was not in caller before (new info from callee is always better)
	- folding over the tainted set, caller is updated lval-wise for every lval that is tainted
- `relationAnalysis.apron`: from the caller only the globals and reachable locals are removed, which are in the tainted set. The resulting state is unified with the callee state (only making the caller state more precise)
- `condVars`: only untainted information is kept. Callee state is not merged.
- `varEq`: only untainted information is kept from caller. Callee state is merged
 

A necessary but noteworthy change was the addition of the `(f_ask : Queries.ask)` to the signature of the combine function, which allows querying the return state of the called function.

Regression tests for each analysis have been added as well as some for specific edge cases.

michael-schwarz commented:
Thank you for the PR! I'll most likely only get around to review it in the first week of January.

michael-schwarz reviewed (approved):
After this small typo is fixed, it looks good to me!

FelixKrayer commented on src/analyses/abortUnless.ml:
```diff
@@ -46,7 +46,7 @@ struct
     [false, candidate]
 
-  let combine ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (au:D.t) : D.t =
+  let combine ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (au:D.t) (f_aks: Queries.ask) : D.t =
```
changed

FelixKrayer commented on src/framework/constraints.ml:
```diff
@@ -615,25 +615,24 @@ struct
+      let f_ask = sync_ctx.ask in
       if M.tracing then M.trace "combine" "function: %a\n" S.D.pretty fd;
-      let r = S.combine {ctx with local = cd} lv e f args fc fd in
+      let r = S.combine {ctx with local = cd} lv e f args fc fd {Queries.f = f_ask} in
```
changed

FelixKrayer commented on src/analyses/base.ml:
```diff
@@ -2252,9 +2285,21 @@ struct
-        let cpa' = CPA.fold CPA.add cpa_noreturn cpa_local in (* add cpa_noreturn to cpa_local *)
-        { fun_st with cpa = cpa' }
+        begin if (Q.LS.is_top tainted) then
+          let cpa_local = CPA.filter (fun x _ -> not (is_global (Analyses.ask_of_ctx ctx) x)) st.cpa in
```
changed here and in other similar places

FelixKrayer commented on src/analyses/base.ml:
```diff
@@ -2241,9 +2241,42 @@ struct
+      else st) tainted_lvs local_st
+
+  let combine ctx (lval: lval option) fexp (f: fundec) (args: exp list) fc (after: D.t) (f_ask: Q.ask) : D.t =
+    let tainted = f_ask.f Q.MayBeTainted in
```
I moved it a bit more inwards. However, I don't think that changes much.
To avoid unconditionally querying one could check if we are context-sensitive or are using partial contexts. Then we could decide by that if we use the old combine style (when completely ctx sensitive) and only send the query if it has the potential to improve the situation (when using partial ctx).
However, I am not sure if this check would be beneficial.

FelixKrayer commented on src/analyses/taintPartialContexts.ml:
```diff
@@ -0,0 +1,110 @@
+
+  let startstate v = D.bot ()
+  let threadenter ctx lval f args = [D.top ()]
+  let threadspawn ctx lval f args fctx = ctx.local
```
This was handled by `special` for the thread create functions. But it is sure better also to address it here.

sim642 replied:
There's a subtlety here, which is why it needs to be handled in both. Namely for
```c
ret = pthread_create(&id, NULL, fun, NULL);
```
In `special` the `lval` is `ret`, which is written some error value. But in `threadspawn` the `lval` is `id`, which the thread ID is written to.

michael-schwarz commented on src/analyses/base.ml:
```diff
@@ -2241,7 +2241,38 @@ struct
+        if M.tracing then M.trace "taintPC" "updating %a; type: %a\n" Lval.CilLval.pretty (v, o) d_type lval_type;
+        match CPA.find_opt v (fun_st.cpa) with
+        | None -> st
+        | Some (`Array _) when (get_string "ana.base.arrays.domain") = "partitioned" -> begin
```
In the latest version, some arrays may be partitioned, while others are not: One would need to check for this specific array here.

FelixKrayer replied:
I have fixed this by adding `domain_of_t` to the signature of `ArrayDomain.S` and using this to check if the Array is partitioned:
`| Some ('Array a) when (CArrays.domain_of_t a) = PartitionedDomain ->`

A different approach I thought of was to use `ArrayDomain.get_domain`:
`(ArrayDomain.get_domain ~varAttr:v.vattr ~typAttr:(typeAttrs lval_type)) = PartitionedDomain`

I chose the first idea as it seemed more straightforward to "ask the array what it is" instead of inspecting the attributes, but I will of course change the approach if you prefer not adding a function to the `ArrayDomain.S` signature.

FelixKrayer commented on src/analyses/base.ml:
```diff
@@ -2241,9 +2241,42 @@ struct
+          match new_arry_opt with
+          | None -> st
+          | Some new_arry -> {st with cpa = CPA.add v new_arry st.cpa}
+          end
```
It would be possible to taint the whole array when a lval inside a partitioned array is written.
However special casing this inside the `taintPartialContext` analysis does not seem like a better approach, as there might be other analyses that can benefit from the finer taint information for partitioned arrays.

Other than that I was not able to come up with a better idea to address this issue.

sim642 reviewed (commented):
Has there been an SV-COMP run with this analysis enabled? Given that a regression test from sv-benchmarks has been adapted, I would guess so?

michael-schwarz commented:
Yes, here they are: [results-felix-all.zip](https://github.com/goblint/analyzer/files/10579453/results-felix-all.zip). 

The first run is: base ctx_insens, no taint
The second one is: base ctc_insens, taint
The third one is: base ctc_sens, no taint

michael-schwarz commented:
@sim642 Would you mind doing your review of this at your earliest convenience? I think I will be able to build on this for the detection if non-volatile locals are modified since a longjmp/setjmp for #970.

sim642 reviewed (approved):
The test group should be changed from 64-taint to 65-taint, because there's 64-noreturn now. Other than that this should be good.

FelixKrayer commented:
I rebased onto the master and renamed the test directory.
Additionally I changed a small thing in the base combine, so that new mappings that are also tainted are not added to the callee state and then updated again in combine_st. Changes nothing in the result but removes unnecessary computations

sim642 commented:
Somehow 65-taint/04-multithread _marshaling_ fails on MacOS: https://github.com/goblint/analyzer/actions/runs/4203078719.
